packages/keymaps/src/browser/keymaps-service.ts

```diff
@@ -0,0 +1,114 @@
+import { FrontendApplicationContribution, FrontendApplication } from '@theia/core/lib/browser';
+import { Disposable, DisposableCollection, CommandRegistry, KeybindingRegistry, ILogger, ResourceProvider } from '@theia/core/lib/common';
+import { RawKeybinding } from '@theia/core/lib/common';
+import { UserStorageUri, UserStorageResource } from '@theia/userstorage/lib/browser/';
+import URI from '@theia/core/lib/common/uri';
+import * as ajv from 'ajv';
+
+export const keymapsSchema = {
+    "type": "array",
+    "properties": {
+        "keybinding": {
+            "type": "string"
+        },
+        "command": {
+            "type": "string"
+        },
+        "context": {
+            "type": "string"
+        },
+    },
+    "required": ["command", "keybinding"],
+    "optional": [
+        "context"
+    ],
+    "additionalProperties": false
+};
+
+export const keymapsUri: URI = new URI('keymaps.json').withScheme(UserStorageUri.SCHEME);
+
+@injectable()
+export class KeymapsService implements Disposable, FrontendApplicationContribution {
+
+    protected readonly toDispose = new DisposableCollection();
+    protected readonly ajv = new ajv();
+    protected ready = false;
+    protected keymapsResource: UserStorageResource;
+
+    constructor(
+        @inject(ResourceProvider) protected readonly resourceProvider: ResourceProvider,
+        @inject(CommandRegistry) protected readonly commandRegistry: CommandRegistry,
+        @inject(KeybindingRegistry) protected readonly keyBindingRegistry: KeybindingRegistry,
+        @inject(ILogger) protected readonly logger: ILogger
+    ) {
+        this.resourceProvider(keymapsUri).then(resource => {
+            if (resource !== undefined && resource instanceof UserStorageResource) {
+                this.keymapsResource = <UserStorageResource>resource;
+                this.toDispose.push(this.keymapsResource);
+                this.keymapsResource.onDidChangeContents(() => {
+                    this.onDidChangeKeymap();
+                });
+
+                this.keymapsResource.readContents().then(content => {
+                    this.parseKeybindings(content);
+                });
+
+            } else {
+                this.logger.error('No resource exists for: ', keymapsUri);
+            }
+        });
+    }
+
+    dispose(): void {
+        this.toDispose.dispose();
+    }
+
+    onStart(app: FrontendApplication): void {
+        this.ready = true;
+    }
+
+    protected onDidChangeKeymap(): void {
+
+        this.keymapsResource.readContents().then(content => {
+            this.parseKeybindings(content);
+        });
+    }
+
+    protected parseKeybindings(content: string) {
+        let keybindings: any;
+        try {
+            keybindings = JSON.parse(content);
+            if (keybindings) {
+                this.setKeymap(keybindings);
+            }
+        } catch (error) {
+            this.keyBindingRegistry.resetKeybindings();
```

Comment:
There were keybindings registered that worked. Now the user changed the keybindings and broke the JSON and as a result we reset the working keybindings? I don't think that should happen.

Reply:
As stated before, I did it this way because VScode does it like that. To me it makes sense to provide zero custom keybindings if your keybindings file is invalid, this way I really know something is wrong with my config.

Reply:
ok, let's go with this behavior and see how it feels in practice.